### ISSUE TITLE
Build fails with plain `make`

### ISSUE THREAD

**Issue by alerque** [pjsip/pjproject#2300]
This project, and especially the swig components, will not build consistently with a plain `make` invocation. It works fine for me with a multi-process parallel build (`make -j6` or similar`) and it works _for me_ with  `make` because my system default MAKEFLAGS specify parallel builds. However it dies for people that don't have this default. Specifically enabling the job server even without doing a parallel build with `make -j1` solves the problem.

This should be fixed so either `make` or `make -jN` works across the board.

See also comments on [Arch Linux User Repository packaging](https://aur.archlinux.org/pkgbase/pjproject/#comment-730535).

**Comment by sauwming** [comment]
Does this only happen in Arch Linux (or can I also replicate this issue on Ubuntu for example)?

Unfortunately, I'm not too familiar with the platform, so if you have any suggestion on how to fix this, please feel free to let us know. A Pull Request will be more than welcome.

**Comment by alerque** [comment]
@sauwming I'm not in a position to test that. I do know a lot of people on Arch had the same problem. I just don't have any other distros handy to check with these days.

I've done a lot of work with GNU-Make and this is the first time I've run into _this_ problem. Usually people's builds break when doing parallel builds because they don't actually have the dependency tree straight and are accidentally relying on things running in the sequential order of the Makefile. I've never seen the opposite like this before!

**Comment by sauwming** [comment]
Can you provide us the error(s) that make gives when it fails?

**Comment by alerque** [comment]
From [this comment](https://aur.archlinux.org/pkgbase/pjproject/#comment-730448), here is a tail at least:

```
make[2] : on quitte le répertoire « /tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/build »
make[2] : on quitte le répertoire « /tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/build »
make[1] : on quitte le répertoire « /tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/build »
make : on entre dans le répertoire « /tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig »
make  -C python 
make[1]: Entering directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python'
python3 setup.py build 
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
running build
running build_py
running build_ext
building '_pjsua2' extension
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -DNDEBUG -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.8 -c pjsua2_wrap.cpp -o build/temp.linux-x86_64-3.8/pjsua2_wrap.o make[2]: Entering directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python' -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -DPJ_AUTOCONF=1 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -DNDEBUG -DPJ_IS_BIG_ENDIAN=0 -DPJ_IS_LITTLE_ENDIAN=1 -fPIC -DPJMEDIA_VIDEO_DEV_HAS_SDL=1 -I/usr/include/SDL2 -D_REENTRANT -DPJMEDIA_HAS_LIBAVDEVICE=1 -DPJMEDIA_HAS_LIBAVFORMAT=1 -DPJMEDIA_HAS_LIBAVCODEC=1 -DPJMEDIA_HAS_LIBSWSCALE=1 -DPJMEDIA_HAS_LIBAVUTIL=1 -DPJMEDIA_VIDEO_DEV_HAS_V4L2=1 -DPJMEDIA_HAS_VPX_CODEC=1 -DPJMEDIA_VIDEO_DEV_HAS_SDL=1 -I/usr/include/SDL2 -D_REENTRANT -DPJMEDIA_HAS_LIBAVDEVICE=1 -DPJMEDIA_HAS_LIBAVFORMAT=1 -DPJMEDIA_HAS_LIBAVCODEC=1 -DPJMEDIA_HAS_LIBSWSCALE=1 -DPJMEDIA_HAS_LIBAVUTIL=1 -DPJMEDIA_VIDEO_DEV_HAS_V4L2=1 -DPJMEDIA_HAS_VPX_CODEC=1 -I/tmp/makepkg/pjproject/src/pjproject-2.10/pjlib/include -I/tmp/makepkg/pjproject/src/pjproject-2.10/pjlib-util/include -I/tmp/makepkg/pjproject/src/pjproject-2.10/pjnath/include -I/tmp/makepkg/pjproject/src/pjproject-2.10/pjmedia/include -I/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip/include -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fPIC -DNDEBUG make[2]: Leaving directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python'
gcc: error: make[2]: Entering directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python': No such file or directory
gcc: error: make[2]: Leaving directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python': No such file or directory
error: command 'gcc' failed with exit status 1
make[1]:  [Makefile:37: _pjsua2.so] Error 1
make[1]: Leaving directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python'
make:  [Makefile:27 : python] Erreur 2
make : on quitte le répertoire « /tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig »
==> ERREUR : Une erreur s’est produite dans build().
    Abandon…
```

I've requested anybody else that can replicate the failure to get you a full dump (maybe even in English :wink:).

**Comment by edacval** [comment]
> Can you provide us the error(s) that make gives when it fails?

[pjproject-2.10-1-x86_64-build.log](https://github.com/pjsip/pjproject/files/4253880/pjproject-2.10-1-x86_64-build.log)

site_config.h:
```
#define PJ_DEBUG 0
#define PJ_TIMER_DEBUG 0
#define PJ_HAS_IPV6 1
```
Used default Arch Linux build env settings: 
```
CARCH="x86_64"
CHOST="x86_64-pc-linux-gnu"
CPPFLAGS="-D_FORTIFY_SOURCE=2"
CFLAGS="-march=x86-64 -mtune=generic -O2 -pipe -fno-plt"
CXXFLAGS="-march=x86-64 -mtune=generic -O2 -pipe -fno-plt"
LDFLAGS="-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now"
```
`make` jobserver settings:
```
MAKEFLAGS="j13" 
```

**Comment by alerque** [comment]
@edacval Are you sure that's your system MAKEFLAGS setting? Shouldn't it start with a dash?

```
MAKEFLAGS="-j13"
```

**Comment by edacval** [comment]
> @edacval Are you sure that's your system MAKEFLAGS setting? Shouldn't it start with a dash?
> 
> ```
> MAKEFLAGS="-j13"
> ```

True . Fixed

**Comment by alerque** [comment]
@edacval It looks like the build log you posted was from `makepkg`. That will be affected by that bogus setting. I assume it probably works for you now (much like mine does with `-j6`). Are you able to replicate a failure manually running make (with no `-j` argument as shown in your steps to reproduce?

**Comment by sauwming** [comment]
@edacval. It looks like a different problem.
I found a typo when checking our build files, so #2306 should be able to fix this issue.

As for the original issue (failure of make with no -j argument), I still have no idea on how to fix it. If available, please provide us another log when when the failure happens during the main pjproject build (the log tail above shows the error when it happens during the python swig build).

**Comment by alerque** [comment]
@sauwming This might only be a problem with the Python SWIG build. I've heard *reports* of it on the main build build but never been able to get them confirmed, I think people's testing is weak because they assume all or nothing and change all the make instances (or none of them). That's not to say there isn't a problem, but the SWIG build is where I've seen the most independent sources for this issue.

**Comment by sauwming** [comment]
Is this the current steps you use to build PJSIP?
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=pjproject
```
make -j1 -C pjsip-apps/src/swig
cd 'pjsip-apps/src/swig/python'
python setup.py build
```

If yes, then I believe there's some redundancy as both will build Python and it may cause an issue if they're executed in parallel.
```
cd 'pjsip-apps/src/swig/'
make python
```
should be enough.

**Comment by alerque** [comment]
Just in case this turns out to be relevant, GNU Make 4.3 was just recently release with [some changes](https://git.savannah.gnu.org/cgit/make.git/tree/NEWS) to the job server. My favorite is you can now specific the number of jobs for the MAKEFLAGS _inside_ a Makefile! [So. Long. Coming.](https://stackoverflow.com/q/42291992/313192).

Arch Linux _does_ have this packaged and distributed already, but I don't *think* it is relevant to this bug. The hack to force `-j1` was setup in the PKGBUILD for pjproject _long_ ago.

@sauwming In answer to your question, no the `make` and `python setup.py` steps are not being done in parallel, first the `build()` function runs (which is where we are running `make`, then later when it is confirmed compelte a `package()` function runs that runs `python setup.py install` (which automatically triggers the build). I can change that around, but there isn't a parallel problem there.

**Comment by sauwming** [comment]
That's great news about the GNU Make! Although yes, it doesn't mean that the issue itself is resolved.

Looking at the error posted above, it seems strange:
```
gcc: error: make[2]: Entering directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python': No such file or directory
gcc: error: make[2]: Leaving directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python': No such file or directory
error: command 'gcc' failed with exit status 1
```
since the directory clearly exists.

I'm not sure how makepkg runs but perhaps it has prematurely deleted pjproject directory. Another thing that I don't know is if the error is always like that. If there's another log, I'll be interested to look.

**Comment by edacval** [comment]
Well, isolated bug cause.
When `make -C pjsip-apps/src/swig -j N>1` is invoked every sub-make job  is set to print working directory to STDOUT by default  so every `make` invocation  in pjsip-apps/src/swig/python/setup.py  completely ignores command line params `--no-print-directory`. This will cause  to invalid variable values assignment from dirty STDOUT in pjsip-apps/src/swig/python/setup.py . For example: 
 [70-72 Lines](https://github.com/edacval/pjproject/blob/072fee1e68175dc7406ab63968514a45aab43236/pjsip-apps/src/swig/python/setup.py#L70) `pj_target_name` will be assigned with
```
make[1]: Entering directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python'                                                                                          
x86_64-unknown-linux-gnu                                                                                                                                                                    
make[1]: Leaving directory '/tmp/makepkg/pjproject/src/pjproject-2.10/pjsip-apps/src/swig/python' 
```
instead of 
```
x86_64-unknown-linux-gnu                                                                                                                                                                    
```
value. Later `setup()` in pjsip-apps/src/swig/python/setup.py tries to  compile extension with invalid `pj_target_name` variable value and  compiler will fail with very misleading message about missing directory.
This can be  100% reproductable observed by applying patch from https://github.com/edacval/pjproject/commit/f9505f4d2b5e3408faebfd03c07c4cf2526a87fa and calling `make -C pjsip-apps/src/swig -j N>1`

How to fix:
1. Quick&fragile way.
Apply any of https://github.com/edacval/pjproject/commit/1cc27a0c4adee56d80c998659c309f0bbe64a6c0 or https://github.com/edacval/pjproject/commit/914bb1d7fdbe7055ef11cb86d253e79ef8279e56 patch. This wil force every `make` invocation  in pjsip-apps/src/swig/python/setup.py to ignore parent `make` environment .
2. Proper way.
Move `make` tasks in pjsip-apps/src/swig/python from setup.py to Makefile. Save tasks output to separate temporary files and tell  setup.py to read values from them. Quick&dirty prototype patch  https://github.com/edacval/pjproject/commit/0a4e7a0bbf3a2dd6aadf67070c79654ffa432cf6 to fix `pj_target_name` variable assignment, needs testing and polishing. Simillar patching is needed for `extra_compile_args`, `libraries` and `extra_link_args` variables in pjsip-apps/src/swig/python/setup.py

That was fun to debug :)

**Comment by alerque** [comment]
Hats off to you sir @edacval, that is not my idea of fun!

**Comment by sauwming** [comment]
Thanks to @edacval for the great work! I applied his patch and managed to reproduce it on a [Ubuntu CI machine here](https://github.com/sauwming/pjproject/runs/474015137?check_suite_focus=true). (note that this Github Action result may disappear in the future).

Although strangely, I can't reproduce it on my local Ubuntu and Mac machines despite following the same steps.

I like the proper solution you proposed, but that seems like a lot of modification to try to get rid of the working directory. So I consulted the [Make documentation](https://www.gnu.org/software/make/manual/html_node/_002dw-Option.html) and tried several options, such as `-s` (which doesn't work), and it seems that adding `--no-print-directory` when invoking the `make` works. Please check https://github.com/sauwming/pjproject/commit/3e4b04e50cc326a2a4c25f0860e2aa6ac5144a97 and let us know if it also works there. In https://github.com/sauwming/pjproject/commit/c1970f4c5ceb7d89243c6a202cabe1104efb7d7c, I also tried to disable the parallel build by moving the '-j1' inside the Makefile and it seems to work too (disabling parallel build should be okay here since basically we will invoke only a single swig command to create the swig binding).

As a side note, this seems to confirm that the issue actually happens when parallel build is invoked, instead of the other way around. Which means that it's probably better to change the issue title to:
Parallel build failure for swig python component.
Unless there's any additional info or log that shows a different error?

**Comment by sauwming** [comment]
FYI, I have combined the patches to fix both (pjsua2 and swig) build issues in #2306. Please let us know if the issue persists.

**Comment by alerque** [comment]
@sauwming Thanks for all your work on this. I'm afraid it's going to be kind of hard to get good evidence it's fixed for everybody until the next release cycle. It was hard for me to replicate the failure case because all my systems have non-default make setups. When I drop the work-around from the distro packaging and push out the next new release I'm sure we'll hear about it if it doesn't work.

That's not to say I'm not trying ... I have tested this and it seems to work for me as best I can remember the setup that was failing, but I would call the testing brittle and incomplete at best right now. I don't have a clear failing on commit X, working on Y. Perhaps @edacval can do better? I'll report back if I get time to test better as well.

**Comment by sauwming** [comment]
Closing this for now as it should have been fixed by #2306. Feel free to reopen the issue (or create a new one) if any problem arises.
